File: sci-matlab/src/sci_matlab/matlab/+scidb/Log.m
classdef Log
%SCIDB.LOG  Minimal structured logger with levels and timestamps.
%
%   scidb.Log.set_level('DEBUG')   — show all messages
%   scidb.Log.set_level('INFO')    — show INFO, WARN, ERROR (default)
%   scidb.Log.set_level('WARN')    — show WARN, ERROR only
%   scidb.Log.set_level('ERROR')   — show ERROR only
%
%   scidb.Log.debug('Processing %d items', n)
%   scidb.Log.info('Loaded %s', type_name)
%   scidb.Log.warn('No data for %s', key)
%   scidb.Log.err('Failed: %s', msg)
%
%   Output format: [HH:MM:SS.FFF] [LEVEL] message
%
%   Writes to scidb.log next to the database file (set automatically
%   by scidb.configure_database()). Each log call opens, appends, and
%   closes the file so every line is flushed to disk immediately.
%
%   Uses setappdata(0, ...) for global state (standard MATLAB singleton pattern).

    properties (Constant)
        DEBUG = 0
        INFO  = 1
        WARN  = 2
        ERROR = 3
    end

    methods (Static)

        function set_level(level)
        %SET_LEVEL  Set the global log level.
        %   Accepts a string ('DEBUG','INFO','WARN','ERROR') or numeric (0-3).
            if ischar(level) || isstring(level)
                level = scidb.Log.parse_level(char(upper(string(level))));
            end
            setappdata(0, 'scidb_log_level', level);
        end

        function level = get_level()
        %GET_LEVEL  Get the current log level (default: INFO).
            if isappdata(0, 'scidb_log_level')
                level = getappdata(0, 'scidb_log_level');
            else
                level = scidb.Log.INFO;
            end
        end

        function set_path(log_path)
        %SET_PATH  Set the log file path for file output.
        %   Called automatically by scidb.configure_database().
            setappdata(0, 'scidb_log_path', char(log_path));
        end

        function p = get_path()
        %GET_PATH  Get the current log file path (empty if not set).
            if isappdata(0, 'scidb_log_path')
                p = getappdata(0, 'scidb_log_path');
            else
                p = '';
            end
        end

        function debug(fmt, varargin)
        %DEBUG  Log a message at DEBUG level.
            if scidb.Log.get_level() <= scidb.Log.DEBUG
                scidb.Log.emit('DEBUG', fmt, varargin{:});
            end
        end

        function info(fmt, varargin)
        %INFO  Log a message at INFO level.
            if scidb.Log.get_level() <= scidb.Log.INFO
                scidb.Log.emit('INFO', fmt, varargin{:});
            end
        end

        function warn(fmt, varargin)
        %WARN  Log a message at WARN level.
            if scidb.Log.get_level() <= scidb.Log.WARN
                scidb.Log.emit('WARN', fmt, varargin{:});
            end
        end

        function err(fmt, varargin)
        %ERR  Log a message at ERROR level.
        %   Named 'err' to avoid conflict with MATLAB's built-in 'error'.
            if scidb.Log.get_level() <= scidb.Log.ERROR
                scidb.Log.emit('ERROR', fmt, varargin{:});
            end
        end

    end

    methods (Static, Access = private)

        function emit(level_str, fmt, varargin)
        %EMIT  Format and write a log message with timestamp to the log file.
        %   Opens, appends, and closes the file on every call so that
        %   each line is flushed to disk immediately.
            log_path = scidb.Log.get_path();
            if isempty(log_path)
                return;
            end

            ts = datestr(now, 'HH:MM:SS.FFF'); %#ok<TNOW1,DATST>
            msg = sprintf(fmt, varargin{:});

            fid = fopen(log_path, 'a');
            if fid ~= -1
                fprintf(fid, '[%s] [%s] %s\n', ts, level_str, msg);
                fclose(fid);
            end
        end

        function level = parse_level(name)
        %PARSE_LEVEL  Convert a level name string to numeric value.
            switch name
                case 'DEBUG'
                    level = scidb.Log.DEBUG;
                case 'INFO'
                    level = scidb.Log.INFO;
                case 'WARN'
                    level = scidb.Log.WARN;
                case 'ERROR'
                    level = scidb.Log.ERROR;
                otherwise
                    warning('scidb:Log', 'Unknown log level ''%s'', defaulting to INFO.', name);
                    level = scidb.Log.INFO;
            end
        end

    end

end

File: scidb/src/scidb/log.py
"""Minimal structured logger for scidb with levels and file output.

Usage:
    from scidb.log import Log

    Log.set_level('DEBUG')          # show all messages
    Log.set_level('INFO')           # show INFO, WARN, ERROR (default)
    Log.set_level('WARN')           # show WARN, ERROR only
    Log.set_level('ERROR')          # show ERROR only

    Log.debug('Processing %d items', n)
    Log.info('Loaded %s', type_name)
    Log.warn('No data for %s', key)
    Log.error('Failed: %s', msg)

Output format: [HH:MM:SS.FFF] [LEVEL] message

Writes to scidb.log next to the database file (set automatically by
configure_database()). Each log call opens, appends, and closes the file
so every line is flushed to disk immediately.
"""

import threading
from datetime import datetime


class Log:
    """Singleton-style logger with file output and log levels."""

    DEBUG = 0
    INFO = 1
    WARN = 2
    ERROR = 3

    _level: int = INFO
    _path: "str | None" = None
    _lock = threading.Lock()

    @classmethod
    def set_level(cls, level: "int | str") -> None:
        """Set the global log level.

        Args:
            level: A string ('DEBUG', 'INFO', 'WARN', 'ERROR') or numeric (0-3).
        """
        if isinstance(level, str):
            level = cls._parse_level(level.upper())
        cls._level = level

    @classmethod
    def get_level(cls) -> int:
        """Get the current log level (default: INFO)."""
        return cls._level

    @classmethod
    def set_path(cls, log_path: "str | None") -> None:
        """Set the log file path for file output.

        Called automatically by configure_database().
        """
        cls._path = log_path

    @classmethod
    def get_path(cls) -> "str | None":
        """Get the current log file path (None if not set)."""
        return cls._path

    @classmethod
    def debug(cls, msg: str) -> None:
        """Log a message at DEBUG level."""
        if cls._level <= cls.DEBUG:
            cls._emit("DEBUG", msg)

    @classmethod
    def info(cls, msg: str) -> None:
        """Log a message at INFO level."""
        if cls._level <= cls.INFO:
            cls._emit("INFO", msg)

    @classmethod
    def warn(cls, msg: str) -> None:
        """Log a message at WARN level."""
        if cls._level <= cls.WARN:
            cls._emit("WARN", msg)

    @classmethod
    def error(cls, msg: str) -> None:
        """Log a message at ERROR level."""
        if cls._level <= cls.ERROR:
            cls._emit("ERROR", msg)

    @classmethod
    def _emit(cls, level_str: str, msg: str) -> None:
        """Format and write a log message with timestamp to the log file."""
        if cls._path is None:
            return
        now = datetime.now()
        ts = now.strftime("%H:%M:%S.") + f"{now.microsecond // 1000:03d}"
        line = f"[{ts}] [{level_str}] {msg}\n"
        with cls._lock:
            try:
                with open(cls._path, "a") as f:
                    f.write(line)
            except OSError:
                pass

    @classmethod
    def _parse_level(cls, name: str) -> int:
        """Convert a level name string to numeric value."""
        levels = {
            "DEBUG": cls.DEBUG,
            "INFO": cls.INFO,
            "WARN": cls.WARN,
            "ERROR": cls.ERROR,
        }
        if name not in levels:
            import warnings
            warnings.warn(
                f"Unknown log level '{name}', defaulting to INFO.",
                UserWarning,
            )
            return cls.INFO
        return levels[name]
